### PR TITLE
Refactor and clean dispatcher related code to support vary size gang

### PR DESCRIPTION
**Pull request by pengzhout**
Previously, dispatcher can only create 1-size or N-size gang, this restricts dispatcher in many ways, This PR aims to support dispatcher to create any size of the gang (with partial or even duplicated segments). To do this, this PR contains a bunch of commits to refactor and cleanup dispatcher/gang code/DTM code. Please refer to each commit message for details.

### CODE REVIEW

**Comment by Eulerizeit** [comment]
@danielgustafsson is this something you needed for the CTE work you are doing?

File: src/backend/cdb/cdbutil.c [hunk 1]
```diff
@@ -1212,14 +1213,21 @@ cdbcomponent_getCdbComponents(bool DNSLookupAsError)
 	else if (cdb_component_dbs->fts_version != ftsVersion)
```
Comment (xiong-gang) [inline src/backend/cdb/cdbutil.c]: How about `else if (cdb_component_dbs->fts_version != ftsVersion && getGxactTwophaseSegments() == NIL && TempNamespaceOidIsValid())`

File: src/backend/cdb/dispatcher/cdbdisp.c [hunk 2]
```diff
@@ -592,3 +557,51 @@ cdbdisp_cleanupAllDispatcherState(void)
...
+AtAbort_DispatcherState(void)
+{
+	if (Gp_role != GP_ROLE_DISPATCH)
+		return;
+
+	CdbResourceOwnerWalker(CurrentResourceOwner, cleanupDispatcherHandle);
+
+	Assert(open_dispatcher_handles == NULL);
+
+	if (TempNamespaceOidIsValid() && cdbcomponent_hasBadWriter())
```
Comment (xiong-gang) [inline src/backend/cdb/dispatcher/cdbdisp.c]: add comments to explain it, please.

File: src/backend/cdb/dispatcher/cdbgang.c [hunk 3]
```diff
@@ -293,31 +293,6 @@ createGang(GangType type, int gang_id, int size, int content)
 	return pCreateGangFunc(type, gang_id, size, content);
 }
 
-/*
- * Test if the connections of the primary writer gang are alive.
- */
```
Comment (xiong-gang) [inline src/backend/cdb/dispatcher/cdbgang.c]: Do you remember why did we have this function at the first place?

File: src/backend/cdb/cdbtm.c [hunk 4]
```diff
@@ -49,6 +49,7 @@
 #include "utils/fmgroids.h"
 #include "utils/sharedsnapshot.h"
 #include "utils/snapmgr.h"
+#include "utils/memutils.h"
```
Comment (asimrp) [inline src/backend/cdb/cdbtm.c]: This commit looks independent enough to be part of a separate PR?  Also, no tests?

File: src/backend/cdb/cdbutil.c (hunk 1 above)
Comment (pengzhout) [reply src/backend/cdb/cdbutil.c]: will consider it, thanks.

File: src/backend/cdb/dispatcher/cdbdisp.c (hunk 2 above)
Comment (pengzhout) [reply src/backend/cdb/dispatcher/cdbdisp.c]: sure.

File: src/backend/cdb/dispatcher/cdbgang.c (hunk 3 above)
Comment (pengzhout) [reply src/backend/cdb/dispatcher/cdbgang.c]: It's originally added to test if the writer gang is alive (like segment be killed by 9) by calling an actual recv() syscall. It's removed because of calling syscall for each segments is heavy, there is no need to do the optimization for such a rare case, the error can bump up when the command is actually sent.

**Comment by pengzhout** [comment]
@asimrp thanks, I will update the PR description for the intention and solution for easier reviews and why only some cases are added.

**Comment by pengzhout** [comment]
As @asimrp suggested, I will close this one and separate it to smaller PRs.
